docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,42 +89,90 @@ The current multisig public keys use amino serialization to generate the address
++ The main arguments are speed and separating from `sha256` which is widely used
+  by miners and could potentially be used to find collisions.
++ The function was in the final round of the 2012 NIST hash function competition.
++ It's well studied with security covered in many academic papers.
++ Faster than `sha2` on non ASICs chipsets.
++ It's getting more traction in other blockchains (Pokadot, Sia, Zcash, ...). Related [zcash discussion](https://github.com/zcash/zcash/issues/706#issuecomment-187807410).
++ It's already widely supported by all major programming languages.
++ Cryptography consulting reviled no argument against `blake2b`
+
+
+### Base Address Algorithm
+
+We start with defining a base algorithm for generating addresses. Notably, it's used for Base Accounts (accounts represented by a single key-pair) addresses. For each Public Key schema we need to have an associated `typ` string, which we will discuss in a section below. `hash` is a cryptographic hash function defined in the previous section.
+
+```go
+const A_LEN = 32
+
+func BaseAddress(typ string, pubkey []byte) []byte {
+    return hash(hash(typ) + pubkey)[:A_LEN]
+}
+```
+
+The `+` is a bytes concatenation, which doesn't use any separator.
 
-### Canonical Address Format
+This algorithm is an outcome after a consulting session with a cryptographer.
+Motivation: this algorithm keeps the address relatively small (length of the `typ` doesn't impact on the length of the final address)
+and it's more secure than [post-hash-prefix-proposal] (with reducing the pubkey hash to 20 bytes, we essentially significantly reduce the address space).
+Moreover the cryptographer motivated the choice to add `typ` in the hash to protect against switch table attack.
 
-We have three types of accounts we would like to create addresses for in the future:
-- regular public key addresses for new signature algorithms (ex. `sr25519`).
-- public key addresses for multisig public keys that don't use amino encoding
-- module accounts: basically any accounts which cannot sign transactions and
-which are managed internally by modules
 
-To address all of these use cases we propose the following basic `AddressHash` function,
-based on the discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694):
+### Composed Account Address Algorithm
+
+We will generalize `BaseAddress` algorithm to define an address for an account which is represented by a set of sub accounts (example: group module accounts, multisig acconts...).
+The address is constructed by recursively creating addresses for the sub accounts, sorting the addresses and composing it into a single address:
 
 ```go
-func AddressHash(prefix string, contents []byte) []byte {
-	preImage := []byte(prefix)
-	if len(contents) != 0 {
-		preImage = append(preImage, 0)
-		preImage = append(preImage, contents...)
-	}
-	return sha256.Sum256(preImage)[:20]
+
+type Acc interface {
+    Typ() string
+    SubAccounts() []Acc
+    ...
+}
```

Comment:
We don't really have a structure like this and I don't think this is how we would do this.

How about just two functions `RootModuleAddress` and `DerivedModuleAddress`?

Reply:
I know. This is just an idea. We want an account to be able to:
+ present it's type (message name)
+ list subaccounts (possible empty - in that case it's reduced to a BaseAccount).

Reply:
This is a design for future accounts (including group module). This is not for a module account derivation mechanism as presented in ADR-033.

Reply:
Let's discuss this live. I think there is some misunderstanding here. I'm not sure what's wrong with the original text other than changing the address length...

Reply:
This section is outdated and removed - as agreed we don't use an _account_ concept here.